pkg/dataplane/server/forwarder.go

```diff
@@ -0,0 +1,112 @@
+package server
+
+import (
+	"io"
+	"net"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	dataBufferSize = 64 * 1024
+)
+
+type forwarder struct {
+	appConn  net.Conn
+	peerConn net.Conn
+	logger   *logrus.Entry
+}
+
+type connDialer struct {
+	c net.Conn
+}
+
+func (cd connDialer) Dial(_, _ string) (net.Conn, error) {
+	return cd.c, nil
+}
+
+func (f *forwarder) peerToApp() error {
+	bufData := make([]byte, dataBufferSize)
+	for {
+		numBytes, err := f.peerConn.Read(bufData)
+		if err != nil {
+			if err != io.EOF { // don't log EOF
+				return err
+			}
+			break
+		}
+		_, err = f.appConn.Write(bufData[:numBytes]) // TODO: track actually written byte count
+		if err != nil {
+			if err != io.EOF { // don't log EOF
+				return err
+			}
+			break
+		}
+	}
+	f.closeConnections()
```

Comment:
consider calling shutdown read/write instead of calling close on the connections. You may end up tearing down connections before all data has been transmitted (i.e., workload->gw closed, but still have some slower sending on gw->peer as it is over the WAN. CloseConnections will shutdown both)

Reply:
Go has a [tcpConn.CloseRead](https://pkg.go.dev/net#TCPConn.Close)

Reply:
I tried moving the closeConnection after wg.wait. The side-effect is that the one way connection doesn't tear down automatically, and for some reason iperf3 doesn't seem to work with that behavior. Hence, I have tried adding a close signal, which will exit the loop in a graceful way and finally close connection. Hence, to exit the loop, we need to add a read deadline so that it doesn't wait on the connection.

Reply:
here's the pseudocode of the logic I had in mind (simplified):
```go
   nb, err := c1.Read(...)
   if err != nil { // done reading from c1
      if nb > 0 { // write last remaining bytes in c1->c2, if any
         c2.Write(...)
      }
      // signal c2 that we're done writing - this will send a FIN and allow upstream to close c2 
      c2.CloseWrite()
      break // done with the read/write loop, signal the wg outside so that connections can be closed when it's done
   }
   // continue normal processing for read/write when no error on read
   nb, err = c2.Write(...) 
   ...
```

Reply:
I did a quick try to use TCPConn. However, Hijacker connection is unable to be cast to TCPConn. Hence it failed to establish connection.

Reply:
Since we started with an HTTP socket, built on top of TCP, the hijacked `net.Conn` connection should be a `net.TCPConn`. I don't see how it would fail the cast/type-check. For debug, can you inspect the type of connection object returned?